src/osx/volume.mm

```diff
@@ -0,0 +1,142 @@
+// Name:        src/osx/volume.cpp
+// Purpose:     apple implementation of wxFSVolume
+// Author:      Tobias Taschner
+// Created:     2021-11-17
+// Copyright:   (c) 2021 wxWidgets development team
+// Licence:     wxWindows licence
+///////////////////////////////////////////////////////////////////////////////
+
+// ============================================================================
+// declarations
+// ============================================================================
+
+// ----------------------------------------------------------------------------
+// headers
+// ----------------------------------------------------------------------------
+
+#include "wx/wxprec.h"
+
+
+#if wxUSE_FSVOLUME && !defined(__WXOSX_IPHONE__)
+
+#include "wx/volume.h"
+
+#ifndef WX_PRECOMP
+    #if wxUSE_GUI
+        #include "wx/icon.h"
+    #endif
+#endif // WX_PRECOMP
+
+#include "wx/arrimpl.cpp"
+
+#include "wx/osx/core/cfref.h"
+#include "wx/osx/core/cfstring.h"
+
+#import <Foundation/NSString.h>
+#import <Foundation/NSFileManager.h>
+
+//+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+// wxFSVolume
+//+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+wxArrayString wxFSVolumeBase::GetVolumes(int flagsSet, int flagsUnset)
+{
+    auto nativeVolumes = [[NSFileManager defaultManager]
+                          mountedVolumeURLsIncludingResourceValuesForKeys:nil
+                          options:NSVolumeEnumerationSkipHiddenVolumes];
+
+    wxArrayString volumePaths;
+    for (NSURL* url in nativeVolumes)
+    {
+        wxFSVolumeBase volume(url.fileSystemRepresentation);
+        int flags = volume.GetFlags();
+        if ((flags & flagsSet) == flagsSet && !(flags & flagsUnset))
+            volumePaths.push_back(volume.GetName());
+    }
+
+    return volumePaths;
+}
+
+void wxFSVolumeBase::CancelSearch()
+{
+}
+
+wxFSVolumeBase::wxFSVolumeBase()
+{
+    m_isOk = false;
+}
+
+wxFSVolumeBase::wxFSVolumeBase(const wxString& name)
+{
+    Create(name);
+}
+
+bool wxFSVolumeBase::Create(const wxString& name)
+{
+    m_isOk = false;
+    m_volName = name;
+
+    NSURL* url = [NSURL fileURLWithPath:wxCFStringRef(name).AsNSString()];
+    auto values = [url resourceValuesForKeys:@[NSURLVolumeLocalizedNameKey] error:nil];
+    if (values)
+    {
+        m_isOk = true;
+        m_dispName = wxCFStringRef::AsString((CFStringRef)[values objectForKey:NSURLVolumeLocalizedNameKey]);
+    }
+
+    return m_isOk;
+}
+
+bool wxFSVolumeBase::IsOk() const
+{
+    return m_isOk;
+}
+
+wxFSVolumeKind wxFSVolumeBase::GetKind() const
+{
+    NSURL* url = [NSURL fileURLWithPath:wxCFStringRef(GetName()).AsNSString()];
+    auto values = [url resourceValuesForKeys:@[NSURLVolumeIsLocalKey] error:nil];
+
+    // Assume disk for local volumes
+    if ([(NSNumber*)[values objectForKey:NSURLVolumeIsLocalKey] boolValue])
+        return wxFS_VOL_DISK;
+    else
+        return wxFS_VOL_NETWORK;
+}
```

Comment:
To make `wxGetAvailableDrives` work as before, this must at least return cdrom for read only drives

```
wxFSVolumeKind wxFSVolumeBase::GetKind() const
{
    NSURL* url = [NSURL fileURLWithPath:wxCFStringRef(GetName()).AsNSString()];
    auto values = [url resourceValuesForKeys:@[NSURLVolumeIsLocalKey, NSURLVolumeIsReadOnlyKey] error:nil];

    // Assume disk for local volumes
    if ([(NSNumber*)[values objectForKey:NSURLVolumeIsLocalKey] boolValue])
    {
        if ([(NSNumber*)[values objectForKey:NSURLVolumeIsReadOnlyKey] boolValue])
            return wxFS_VOL_CDROM;
        else
            return wxFS_VOL_DISK;
    }
    else
        return wxFS_VOL_NETWORK;
}

```